feat: Simple Passkey AA contract version 1

oneleo opened the pull request:
## Simple Passkey AA contract version 1
1. Include the typesetting file: .prettierrc.json, .solhint.json.
2. Set the optimizer runs for P256 verifier in foundry.toml.
3. Add a simple foundry test for Passkey contract: `% forge test -vvv`.

## Foundry Test
```shell
% forge test -vvv
```

## Start Geth Testnet
```
% make testnet-up

testnet-geth_setup-1     | [Deploy EntryPoint]
testnet-geth_setup-1     | Deployed to: 0x5FbDB2315678afecb367f032d93F642f64180aa3
testnet-geth_setup-1     | Transaction hash: 0xb9d8b94e3b8bd9690f13c9ad7cc3e64d4282954684c01f934594829a80ff5a51

testnet-geth_setup-1     | [Deploy SimpleAccountFactory]
testnet-geth_setup-1     | Deployed to: 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
testnet-geth_setup-1     | Transaction hash: 0x00899a4493a6f705405b9b6032fc702879fbd8cf330494422a51cd873048c7bc

testnet-geth_setup-1     | [Create SimpleAccount]
testnet-geth_setup-1     | [Get SimpleAccount address]
testnet-geth_setup-1     | 0x1cee485cc83c5a17692904ff441a115fb223788e
testnet-geth_setup-1     | [Transfer 100 ETH to SimpleAccount]

testnet-geth_setup-1     | [Deploy Counter]
testnet-geth_setup-1     | Deployed to: 0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9
testnet-geth_setup-1     | Transaction hash: 0x673da51fd03cfb77ed8e76f4174d41c0715cef795e3aba9f06b6c32fde2b2675

testnet-geth_setup-1     | [Deploy PasskeyAccountFactory]
testnet-geth_setup-1     | Deployed to: 0x5FC8d32690cc91D4c39d9d3abcBD16989F875707
testnet-geth_setup-1     | Transaction hash: 0xcbc0f2c3e678e7b890812f5ff596aa2b5b3d33fb3410df840db651b293ddc1ad

testnet-geth_setup-1     | [Create PasskeyAccount]
testnet-geth_setup-1     | [Get PasskeyAccount address]
testnet-geth_setup-1     | 0x3c93bc637f7630b2d4ad51f4a02b786e81ff1498
testnet-geth_setup-1     | [Transfer 100 ETH to PasskeyAccount]
```

oneleo committed:
feat: Simple Passkey AA contract version 1

1. Include the typesetting file: .prettierrc.json, .solhint.json.
2. Set the optimizer runs for P256 verifier in foundry.toml.
3. Add a simple foundry test for Passkey contract: `% forge test -vvv`.

oneleo committed:
Contract adds 'credId' parameter as input for Passkey's credential id
1. Provide the 'credId' parameter when creating the PasskeyAccount contract.
2. Format contract code with 'forge fmt'.
3. Revert Foundry's 'solc_version' to 0.8.23.

oneleo committed:
feat:Add initialization script content in the rundler container:

1. Deploy PasskeyAccountFactory contract.
2. Create PasskeyAccount contract.
3. Transfer ETH to PasskeyAccount.

oneleo committed:
chg: Adjust Foundry's formatting and format *.sol files

1. Adjust Foundry's formatting settings.
2. Format *.sol files in this project.

oneleo committed:
chg: Integrate Bitwarden-generated WebAuthn Private Key

1. Stabilize PasskeyAccount address with Bitwarden's WebAuthn
Private Key.
2. Align PubKeyX, PubKeyY, authenticatorData, clientDataJson, r,
and s values in Foundry Test with the WebAuthn Private Key.
3. Update contract address in Readme.md.
4. Use CredentialId, PubKeyX, and PubKeyY in the Container script
for PasskeyAccount deployment.
5. Highlight each Container script step in yellow.

oneleo committed:
data: Set env var in entry.sh, add Bitwarden reg resp, update addr in Readme

1. Set common environment variables (rpc_url, owner_private_key,
simple_account_address, passkey_account_address) to streamline the
entry.sh script.
2. Add the output response file for registering Bitwarden Passkey.
3. Modify the credentialId used when generating the Passkey.
4. Update the SimpleAccount address modified in Readme.md due to
differences in Bytecode, as Compiler Optimizer is used.
5. Update the SimpleAccount address modified in Readme.md due to
different credentialId.

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,260 @@
+        require(msg.sender == address(this), "Only wallet can update passkeys");
+        require(pubKeyX != 0 && pubKeyY != 0, "Zero passkey is not allowed");
+        require(
+            passkey.pubKeyX != 0 && passkey.pubKeyY != 0,
```
nit: I saw many places checking this. Maybe it is worth a util function like `isPasskeySet`.

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,260 @@
+
+contract PasskeyAccount is SimpleAccount, IPasskeyAccount {
+    P256Verifier public immutable p256Verifier;
+    Passkey private passkey;
```
This should allow to read, we could make it public or add another external function `getPasskey`.

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,260 @@
+        emit PasskeyUpdated(credId, passkey.pubKeyX, passkey.pubKeyY);
+    }
+
+    function validateSignature(
+        UserOperation calldata userOp,
+        bytes32 userOpHash // As Webauthn's challenge
+    ) external view returns (uint256) {
+        return _validateSignature(userOp, userOpHash);
+    }
```
Q: Is this function helping us to test? Inheriting SimpleAccount we only need to override `_validateSignature`.

oneleo replied:
Since `_validateSignature()` inherits from `SimpleAccount`, it is restricted to internal. To enable its use in Foundry Test (or potential future applications for advanced signature verification), an external `validateSignature()` method needs to be made accessible.

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,260 @@
+        uint256 y
+    ) internal view returns (bool) {
+        // check for signature malleability
+        if (s > P256_N_DIV_2) {
+            return false;
+        }
```
Q: Shall we alter the `s` when it is greater than `P256_N_DIV_2`, or signature from webauthn might fail irregularly?

pilagod replied:
Or we would enforce client to do this?

oneleo replied:
It should be mandated for the client to follow this approach. If the condition 'if(s>n/2) n-s' is hard-coded in the contract, and a malicious value of s is provided by the client, this hard-coded logic will transform s into a verifiably passable form.

pilagod commented on src/account/PasskeyAccountFactory.sol:
```diff
@@ -0,0 +1,78 @@
+contract PasskeyAccountFactory {
+    PasskeyAccount public immutable accountImplementation;
+
+    constructor(IEntryPoint entryPoint) {
```
nit: Also store entry point on factory and make it queryable

pilagod commented on test/PasskeyAccount.t.sol:
```diff
@@ -0,0 +1,143 @@
+contract PasskeyAccountTest is Test {
+    using UserOperationLib for UserOperation;
+
+    // Get the entryPoint instance from the local node
+    IEntryPoint public entryPoint =
+        IEntryPoint(0x5FbDB2315678afecb367f032d93F642f64180aa3);
```
You can directly deploy a new one here.

pilagod commented on test/PasskeyAccount.t.sol:
```diff
@@ -0,0 +1,143 @@
+    // IEntryPoint public entryPoint =
+    //     IEntryPoint(0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789);
+
+    string constant testCredId = "9h5F3DgLSjSMdnVOadmhCw";
+    uint256 constant testPubKeyX =
+        67299174900712686363169673082376821529726602378544032702281553676098545184711;
+    uint256 constant testPubKeyY =
+        104273800132786176334597151467609377740095818152192999025225464410568038480397;
+    uint256 constant testSalt = 0;
```
nit: The `test` prefix has no much meaning. It would be better to remove it

```suggestion
    string constant credId = "9h5F3DgLSjSMdnVOadmhCw";
    uint256 constant pubKeyX =
        67299174900712686363169673082376821529726602378544032702281553676098545184711;
    uint256 constant pubKeyY =
        104273800132786176334597151467609377740095818152192999025225464410568038480397;
    uint256 constant salt = 0;
```

pilagod commented on test/PasskeyAccount.t.sol:
```diff
@@ -0,0 +1,143 @@
+        uint256 sigS =
+            60564441634584281332486487842777985011692936482102942812772457824444148629022;
+
+        // Do not use abi.encodePacked() here.
```
```suggestion
```

pilagod commented on test/PasskeyAccount.t.sol:
```diff
@@ -0,0 +1,143 @@
+        return this._getUserOpHash(userOp, address(entryPoint));
+    }
+
+    function getUserOpHash(
+        UserOperation memory userOp,
+        address entryPointAddr
+    ) public view returns (bytes32) {
+        return this._getUserOpHash(userOp, entryPointAddr);
+    }
```
Not used

pilagod commented on test/keystore/registrationResponse.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Rename this file to `registration_response.json` to align with the bitwarden key

oneleo committed:
data: Convert credId from Bytes32 to its original string format

1. To restrict the range of credId for Webauthn, it's necessary to
obtain a list of the original credIds.
2. Accordingly, add an exportable parameter 'KnownCredId' to the
contract.

oneleo committed:
data: Modifications to PasskeyAccount-related contracts

1. Altered private to public for non-security-critical variables,
allowing easier user retrieval.
2. Encapsulated frequently used functions into subroutines.
3. Ensured a fixed EntryPoint address in the Foundry Test. Without
it, a new signing with Passkey is necessary for different UserOpHash values.
4. Renamed registrationResponse.json to align with private key
filenames, now regiatration_response.json.

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    function updatePasskey(
+        string calldata credId,
+        uint256 pubKeyX,
+        uint256 pubKeyY
+    ) external virtual override {
+        require(msg.sender == address(this), "Only wallet can update passkeys");
+        require(pubKeyX != 0 && pubKeyY != 0, "Zero passkey is not allowed");
```
What about this refactoring?

```suggestion
        Passkey calldata _passkey
    ) external virtual override {
        require(msg.sender == address(this), "Only wallet can update passkeys");
        require(_isPasskeySet(_passkey), "Zero passkey is not allowed");
```

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,266 @@
+        uint256 pubKeyY
+    ) internal {
+        require(
+            passkey.pubKeyX == 0 && passkey.pubKeyY == 0,
```
```suggestion
            _isPasskeySet(passkey),
```

pilagod commented on src/account/PasskeyAccount.sol:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    // TODO: Assign Passkey with credId.
+    function _isPasskeySet(Passkey memory _passkey)
```
nit: Maybe call it `_isPasskeyValid`

pilagod commented on test/PasskeyAccount.t.sol:
```diff
@@ -0,0 +1,168 @@
+        // Align the entryPoint address with the one on the Gethnode and Mainnet
+        bytes memory entryPointBytecode =
+            vm.getCode("EntryPoint.sol:EntryPoint");
+        address entryPointAnotherAddr =
+            Create2.deploy(0, bytes32(uint256(2)), entryPointBytecode);
+        vm.etch(entryPointGethAddr, entryPointAnotherAddr.code);
+        vm.etch(entryPointMainnetAddr, entryPointAnotherAddr.code);
```
Maybe we don't need `entryPointAnotherAddr`? The second argument of `vm.etch` is bytecode, you may try to use `entryPointBytecode` directly

pilagod replied:
And just choose the geth address (and rename it to `entryPointAddr`), we don't need to be aware chain information in test.

oneleo committed:
data: Delete unused parameters.

oneleo committed:
data: Optimize code

1. Remove redundant import packages.
2. Refactor updatePasskey() to accept input from the Passkey struct.
3. Maximize the use of subroutines.
